drivers/flash/flash_stm32_qspi.c

```diff
@@ -132,38 +137,57 @@ static inline void qspi_set_address_size(const struct device *dev,
 	cmd->AddressSize = QSPI_ADDRESS_24_BITS;
 }
 
-static inline void qspi_prepare_quad_read(const struct device *dev,
+static inline int qspi_prepare_quad_read(const struct device *dev,
 					  QSPI_CommandTypeDef *cmd)
 {
 	struct flash_stm32_qspi_data *dev_data = dev->data;
 
-	if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) && dev_data->flag_quad_io_en) {
+	if (dev_data->flag_quad_io_en) {
+		switch (dev_data->mode) {
+		case JESD216_MODE_114:
+			cmd->AddressMode = QSPI_ADDRESS_1_LINE;
+			cmd->DataMode = QSPI_DATA_4_LINES;
+			break;
+		case JESD216_MODE_144:
+			cmd->AddressMode = QSPI_ADDRESS_4_LINES;
+			cmd->DataMode = QSPI_DATA_4_LINES;
+			break;
+		default:
+			return -ENOTSUP;
+		}
+
 		cmd->Instruction = dev_data->qspi_read_cmd;
-		cmd->AddressMode = QSPI_ADDRESS_4_LINES;
-		cmd->DataMode = QSPI_DATA_4_LINES;
 		cmd->DummyCycles = dev_data->qspi_read_cmd_latency;
 	}
+
+	return 0;
 }
 
-static inline void qspi_prepare_quad_program(const struct device *dev,
+static inline int qspi_prepare_quad_program(const struct device *dev,
 					     QSPI_CommandTypeDef *cmd)
 {
 	struct flash_stm32_qspi_data *dev_data = dev->data;
-	/*
-	 * There is no info about PP/4PP command in the SFDP tables,
-	 * hence it has been assumed that NOR flash memory supporting
-	 * 1-4-4 mode also would support fast page programming.
-	 */
-	if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) && dev_data->flag_quad_io_en) {
-		cmd->Instruction = SPI_NOR_CMD_4PP;
-		cmd->AddressMode = QSPI_ADDRESS_4_LINES;
-		cmd->DataMode = QSPI_DATA_4_LINES;
-		/*
-		 * Dummy cycles are not required for 4PP command -
-		 * data to be programmed are sent just after address.
-		 */
-		cmd->DummyCycles = 0;
+
+	if (dev_data->flag_quad_io_en) {
+		cmd->Instruction = dev_data->qspi_write_cmd;
+
+		switch (cmd->Instruction) {
+		case SPI_NOR_CMD_PP_1_1_4:
+			cmd->AddressMode = QSPI_ADDRESS_1_LINE;
+			cmd->DataMode = QSPI_DATA_4_LINES;
+			cmd->DummyCycles = 0;
```

Comment:
``
			cmd->DataMode = QSPI_DATA_4_LINES;
			cmd->DummyCycles = 0;
``
Why not factorizing these lines ?